### PR TITLE
Fix dev versioning system to allow test PyPI releases

### LINKED ISSUE #500: Fix test PyPI automation
Currently we have a GitHub workflow that uploads to test PyPI on every commit; and to real PyPI on the release. The idea of the test PyPI automation is to check the automation is working and that the package is accepted by PyPI on each commit. In particular this'll catch things like `setup.py` depending on things not in PyPI.

However, this automation is sometimes failing because of version number conflicts. This is because we're automatically generating versions with `setuptools_scm`, which for pre-release appends a `.devN` number. This seems to be guessing the N based on how many commits relative to the last release. But a linear number N that's incremented can't handle there being multiple branches. Usually setuptools_scm handles this with a local scheme including things like a GitHub commit hash -- but test PyPI doesn't accept that, so we've disabled it with `no-local-version`.

I think we either need to remove the test PyPI automation (simplest, but loses some checks), or hack `setuptools_scm` to include the local scheme but in a way that PyPI accepts (e.g. change `-` to `_`).

I'm surprised there isn't a good general solution to this. Some relevant issues: https://github.com/pypa/setuptools_scm/issues/455 and https://github.com/pypa/setuptools_scm/issues/342

### PR DESCRIPTION
**Pull request by Rocamonde**
Fixes #500.

### CODE REVIEW

**Comment by Rocamonde** [comment]
Hooray! https://test.pypi.org/project/imitation/0.3.2.dev30933807/

**Comment by Rocamonde** [comment]
Damn, the `setuptools_scm` library needs to be imported during setup, but the function is defined at the `setup.py` module level, so it gets imported before the installer installs the library, giving an error. Seems overkill to add this as a dependency

**Comment by AdamGleave** [comment]
Is it just a type error? If so could add `setuptools_scm` as a test-only dependency.

**Comment by Rocamonde** [comment]
Actually it's also needed to call the default version guessing script, even though we could just copy this manually from source... This library feels poorly designed.

**Comment by Rocamonde** [comment]
Oh wait, I can just only import the library if we are type checking, then import it within the function scope.

File: setup.py [hunk 1]
```diff
@@ -93,12 +98,34 @@ def run(self):
             )
 
 
+def get_version(version: "ScmVersion") -> str:
+    from setuptools_scm import version as scm_version
```
Comment (AdamGleave) [inline setup.py:102]: We normally avoid imports inside methods. You have a good reason to do this, but can you add a comment to explain why?

File: setup.py [hunk 2]
```diff
@@ -93,12 +98,34 @@ def run(self):
             )
 
 
+def get_version(version: "ScmVersion") -> str:
```
Comment (AdamGleave) [inline setup.py:101]: Add docstring explaining format of version?

File: setup.py [hunk 3]
```diff
@@ -93,12 +98,34 @@ def run(self):
...
+    if version.exact:
+        return version.format_with("{tag}")
+    else:
+        return version.format_next_version(
+            scm_version.guess_next_version,
+            fmt="{guessed}.dev{node}",
+        )
+
+
+def get_local_version(version: "ScmVersion", time_format="%Y%m%d") -> str:
```
Comment (AdamGleave) [inline setup.py:115]: Add docstring explaining what this does. (Was non-obvious to me without looking up format_choice; I see this is something added if working directory is dirty.)

File: setup.py [hunk 4]
```diff
@@ -93,12 +98,34 @@ def run(self):
...
+        "",
+        "+d{time:{time_format}}",
+        time_format=time_format,
+    )
+
+
 setup(
     cmdclass={"install": InstallCommand},
     name="imitation",
     # Disable local scheme to allow uploads to Test PyPI.
```
Comment (AdamGleave) [inline setup.py:126]: This comment is no longer accurate!

**Review by AdamGleave — CHANGES_REQUESTED** [review]
Thanks for fixing this! Happy with the solution at a high-level, few minor requested changes.

File: setup.py [hunk 5]
```diff
@@ -93,12 +98,34 @@ def run(self):
             )
 
 
+def get_version(version: "ScmVersion") -> str:
+    from setuptools_scm import version as scm_version
+
+    if version.node:
+        version.node = str(int(version.node.lstrip("g"), 16))
```
Comment (AdamGleave) [inline setup.py:105]: Where's the g coming from (G for git?)

File: pyproject.toml [hunk 6]
```diff
@@ -2,10 +2,10 @@
 requires = ["setuptools>=45", "setuptools_scm[toml]>=6.2"]
 build-backend = "setuptools.build_meta"
 
-[tools.setuptools_scm]
-# Disable local scheme to allow uploads to Test PyPI.
-# See https://github.com/pypa/setuptools_scm/issues/342
-local_scheme = "no-local-version"
+# [tool.setuptools_scm]
```
[the remark this answers is not in the document]
Comment (Rocamonde) [reply pyproject.toml:5]: Yes, that's right. We should fix this at some point, but I don't see this happening in the near future unless the setuptools_scm decided to support our feature request.

File: setup.py (hunk 5 above)
Comment (Rocamonde) [reply setup.py:105]: Yes, it appends "g" before the first few characters of the commit hash if it comes from git (and does something else if using mercurial, ugh).

File: setup.py (hunk 3 above)
Comment (Rocamonde) [reply setup.py:115]: Apologies, will document. Spent so long reading the docs for this thing that forgot other people don't actually know how this works 😂

**Comment by AdamGleave** [comment]
Linter is complaining about docstring format -- ping me when that's fixed and I'll do another review.

**Comment by Rocamonde** [comment]
Fixed the linting issues, but now getting a separate error which is certainly unrelated as I have not touched code that affects tests in this way:

```TypeError: run() got an unexpected keyword argument 'loggers'```

on `./imitation/scripts/parallel.py:120`:


```python
ray.tune.run(
                trainable,
                config=search_space,
                name=run_name,
                local_dir=local_dir,
                loggers=ray_loggers,
                resources_per_trial=resources_per_trial,
                sync_config=ray.tune.syncer.SyncConfig(upload_dir=upload_dir),
            )
```

**Comment by Rocamonde** [comment]
Yes, they just released ray 2.0.0. 10 hours ago. I'll pin the version and create an issue.

**Comment by Rocamonde** [comment]
Getting these errors on CircleCI which do not show up on my linting:

```
setup.py:147:1: DAR201 Missing "Returns" in Docstring: - return
setup.py:147:1: DAR101 Missing parameter(s) in Docstring: - time_format
setup.py:147:1: DAR101 Missing parameter(s) in Docstring: - version
```

Will fix these now, but any thoughts on the reason behind inconsistent error reporting?

**Comment by Rocamonde** [comment]
Getting errors on CircleCI that are noticed inline by vscode but that do not get pointed out when running `./ci/code_checks.sh`. Any thoughts on why?

```
setup.py:147:1: DAR201 Missing "Returns" in Docstring: - return
setup.py:147:1: DAR101 Missing parameter(s) in Docstring: - time_format
setup.py:147:1: DAR101 Missing parameter(s) in Docstring: - version
```

**Comment by AdamGleave** [comment]
> Getting errors on CircleCI that are noticed inline by vscode but that do not get pointed out when running `./ci/code_checks.sh`. Any thoughts on why? This is a pretty annoying issue, as I can't really go through every file visually, plus it breaks the principle of having local CI checks before pushing.
> 
> ```
> setup.py:147:1: DAR201 Missing "Returns" in Docstring: - return
> setup.py:147:1: DAR101 Missing parameter(s) in Docstring: - time_format
> setup.py:147:1: DAR101 Missing parameter(s) in Docstring: - version
> ```

Do you have `flake8-darglint` installed? If you run `flake8 setup.py` directly does it replicate the error?

The only difference I'm aware of between CI and ci/code_checks.sh is that we have some logic in `ci/code_checks.sh` to only run darglint on files that were changed in a commit (as running darglint on everything is very slow; OK on CI, but not really on each commit hook).

**Review by AdamGleave — APPROVED** [review]
LGTM. I made some cosmetic changes so please review before merging.

**Pull request merged**